Always show the tour on new blogs

roo2 opened the pull request:
Experiment with showing the welcome tour on any blog created in the last 30 minutes (30*60 seconds)
fixes https://github.com/Automattic/wp-calypso/issues/52049

### Testing instructions
`yarn dev --sync` or `install-plugin.sh etk add/always-show-welcome-tour-on-new-blogs` the changes to your sandbox
Sandbox the name that will be used for your new blog e.g. `yourtestsite.wordpress.com`
Force reconnect the a8c  proxy for the sandbox changes to take effect
**Test Part 1**
create a site with the name that you sandboxed (`yourtestsite.wordpress.com`)
the tour should show up again
dismiss the tour and refresh the page
the fact that the tour has been closed is saved to local storage so the tour will not be loaded again
**Test Part 2**
navigate to an existing blog  and open the editor 
the tour should not be shown
**Test Part 3**
wait 30 minutes and test again on `yourtestsite.wordpress.com`
clear local storage for `yourtestsite.wordpress.com`
reload the page
the tour should not be shown

roo2 committed:
show the tour on new blogs

autumnfjeld commented:
Verified:
As an existing user
* I created a new site and the Tour appeared
* I dismissed the tour & refreshed the page and the Tour did not appear, as expected
* I navigated to another existing site and the Tour did not appear, as expected
* I went back to my new site, cleared local storage, and the Tour did not show, as expected

So this works in this test.  I'd like to see what others think about this time based approach, it seems super simple....but am I being too simple? :). Also I don't have a good sense of "blog age", is 5 min appropriate? Is 1 minute better?  🤷

roo2 committed:
set new site threshold to 30 minutes to be in line with the time used in JS

roo2 commented:
I updated the age to be 30 minutes to match what we do in JS https://github.com/Automattic/wp-calypso/blob/38c37bbfe35b8c741c4e346637987ddc0fb3ab59/client/state/sites/selectors/is-new-site.js#L16 it sounds to me like the blog age is a reasonable approach after discussion pbxlJb-Ly-p2

roo2 committed:
remove self-evident comment

roo2 committed:
move the age check into the get_nus_status method so that update_nux_status return result is consistent

roo2 commented:
I moved the check into the `get_nux_status` method so that the logic is centralized and we're not messing with the behavior of `update_nux_status`, I also removed the `FORCE_SHOW_WPCOM_WELCOME_GUIDE` check because it seems to be unused by the code and we probably don't need it for testing any more (?)

cc @autumnfjeld

autumnfjeld commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/class-wp-rest-wpcom-block-editor-nux-status-controller.php:
```diff
@@ -11,6 +11,8 @@
  * Class WP_REST_WPCOM_Block_Editor_NUX_Status_Controller.
  */
 class WP_REST_WPCOM_Block_Editor_NUX_Status_Controller extends \WP_REST_Controller {
+	const NEW_SITE_AGE_SECONDS = 30 * 60;
```
Maybe a comment here about the reasoning for choosing 30 minutes, e.g. for the /start flow where people might need time before they enter the editor,   and maybe link to P2 discussion (if that is normal for including in comments?)

roo2 replied:
I added a comment justification, yes I see we've linked to p2s in other places as well 👍 

I also improved some other comments in the code

autumnfjeld commented:
> I moved the check into the `get_nux_status` method so that the logic is centralized and we're not messing with the behavior of `update_nux_status`, I also removed the `FORCE_SHOW_WPCOM_WELCOME_GUIDE` check because it seems to be unused by the code and we probably don't need it for testing any more (?)
> 
> cc @autumnfjeld

> removed the `FORCE_SHOW_WPCOM_WELCOME_GUIDE` check

Yep, I think its ok to remove `FORCE_SHOW_WPCOM_WELCOME_GUIDE` , I used that flag for dev purposes so I could force the tour on a local wp installation.  @p-jackson might have an opinion.

> moved the check into the `get_nux_status` method so that the logic is centralized 
This makes sense to me.
Should we change the name of this method to be more accurate? `get_show_welcome_guide_status` or something like that?


I'm a little nervous about not having tests for this, but I guess with thorough human testing we might be ok. Just be sure to heavily comment the logic reasoning so if another dev comes along they understand what the rules are.  Maybe a version of. your table or some pseudo code or a link to the P2.

Just curious: how much effort would it be to add php tests, since you said there aren't php tests in this area of the code? I assume you've come to the conclusion the effort is very big compared to the benefit. 😄

p-jackson commented:
> removed the `FORCE_SHOW_WPCOM_WELCOME_GUIDE` check

Fine with me

ETK does have some php unit tests already, like these [coming soon ones](https://github.com/Automattic/wp-calypso/blob/1e0114c9c2f9411a56c3cdb009dcb0b87be1b991/apps/editing-toolkit/editing-toolkit-plugin/coming-soon/test/class-coming-soon-test.php#L21). Looks like you'd just need to [configure it using this file](https://github.com/Automattic/wp-calypso/blob/1e0114c9c2f9411a56c3cdb009dcb0b87be1b991/apps/editing-toolkit/editing-toolkit-plugin/phpunit.xml.dist#L27). I believe they're run locally using `wp-env`.

> any blog created in the last 30 minutes

What I don't necessarily like about this is that it's going to show for users who have explicitly dismissed the tour. So lets say a user has just signed up, created a site and dismissed the tour. Then we deploy this change. We'll end up showing the tour to them again. I feel like it should be "blog created in the last 30 minutes && tour hasn't been explicitly dismissed"

~~Currently the return result from the `wpcom_block_editor_nux_get_status` filter only returns `'dismissed'` or `'enabled'`. But I think we'd want to change that to `'dismissed'`, `'enabled'` or `null`. That way the ETK can decide to treat "enabled" differently from "never seen before". I think only when ...~~

Ok I just realised that this doesn't work and what I really want is a per-site setting. Which is what we were trying to avoid using the simple time method. Maybe a shorter timer is more appropriate ... sorry this has probably all been discussed already on the p2 which I haven't been following

@roo2

p-jackson commented:
Got some ideas from pbxlJb-Ly-p2

Even though the enabled/dismissed state is saved on a per-user basis, it's cached in local storage on a _per-site basis_! That means we're getting a sort of per-site setting that should help minimise situations where users are re-shown the tour right after they've dismissed it.

So what I'm thinking is if we move the logic for using the blog age out of PHP, and move it to JS instead.

So in JS, before we load the enabled/dismissed status into the data store we check if the blog is young, and if it is then we save "enabled" to the data store. (We still need to make the request because we need to know what variant to display).
If the value has been cached in local storage then we don't run any of that code, which means if the user has manually dismissed the tour recently on that site then we won't show the tour.

Apologies, that's sort of tricky to describe in English.

@roo2

p-jackson commented:
Something like this: https://github.com/Automattic/wp-calypso/pull/52402/files

It just ignores the response from the server if `isSiteNew` is `true`. And `fetchWelcomeGuideStatus` is never called if the response has been cached https://github.com/Automattic/wp-calypso/blob/1e0114c9c2f9411a56c3cdb009dcb0b87be1b991/apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/block-editor-nux.js#L35

The only other part is figuring out how to get the site age and passing the `isSiteNew` flag to `fetchWelcomeGuideStatus`. I see it's already available in the Calypso store, you'd just have to figure out how to get it in GB 😄

roo2 commented:
Thanks @p-jackson, I think that moving this logic into JS should be clearer and then we can use the existing isSiteNew check. I'll work on an implementation of that. 👍

roo2 commented:
@p-jackson as you noticed before, it doesn't look like I have easy access to the site creation time from js in the ETK. 

It comes from the site options, which we have easy access to in calypso via the `mapStateToProps` function.
We do use site options in the ETK already, only in the `editing-toolkit-plugin/dotcom-fse` subproject via the [withSiteOptions function](https://github.com/Automattic/wp-calypso/blob/trunk/apps/editing-toolkit/editing-toolkit-plugin/dotcom-fse/lib/site-options/with-site-options.js).

I think if I use this function I should move it out of the `dotcom-fse` subproject, and then make my own getter for the site age (`created_at`) site option.

Do you still think it's worth implementing this in JavaScript? I now feel like the php implementation is quite a bit simpler

p-jackson commented:
Good find @roo2! Nice that there's already something existing in ETK.

> Do you still think it's worth implementing this in JavaScript? I now feel like the php implementation is quite a bit simpler

IMO it is, because I worry about the reports we'll get about the the tour appearing for folks it shouldn't appear for, and then it could be exasperating to track down because it'll be based on timing.

I also agree that the PHP implementation is simpler :) It's relying on a side effect of caching to give us a pseudo per-site setting. Which isn't that clear. So if you wanted to do it "the right way" you could go whole hog and implement a real per-site setting on the backend? This comment pbxlJb-Ly-p2#comment-1018 seemed to take that view.

autumnfjeld commented:
> Good find @roo2! Nice that there's already something existing in ETK.
> 
> > Do you still think it's worth implementing this in JavaScript? I now feel like the php implementation is quite a bit simpler
> 
> IMO it is, because I worry about the reports we'll get about the the tour appearing for folks it shouldn't appear for, and then it could be exasperating to track down because it'll be based on timing.

This is probably ok for the first pass, because it would be very limited anyway. How many times would a user repeatedly enter the editor within the first 30 min? 

> 
> I also agree that the PHP implementation is simpler :) It's relying on a side effect of caching to give us a pseudo per-site setting. Which isn't that clear. So if you wanted to do it "the right way" you could go whole hog and implement a real per-site setting on the backend? This comment pbxlJb-Ly-p2#comment-1018 seemed to take that view.

This might be good if we want to consider future implementations, more granular control, could you estimate the work (a couple days, a week)), so we can decide based on that?  We want to keep this contained.

p-jackson commented:
> How many times would a user repeatedly enter the editor within the first 30 min?

I mean, a lot hopefully? I feel like we'd love for the launch time to be under 30 minutes. So if they're busy popping in and out of the editor while setting up DNS and switching themes before they launch, yeah I think we should expect and hope they do it multiple times.

autumnfjeld commented:
Who is this person that could launch a site in 30 minutes? 😁 I’ve never
understood that.  Just to get a sense of the UI , let alone get your
content in would take hours.

Good to get your insight.  😁

On Thu, Apr 29, 2021 at 15:59 Philip Jackson ***@***.***>
wrote:

> How many times would a user repeatedly enter the editor within the first
> 30 min?
>
> I mean, a lot hopefully? I feel like we'd love for the launch time to be
> under 30 minutes. So if they're busy popping in and out of the editor while
> setting up DNS and switching themes before they launch, yeah I think we
> should expect and hope they do it multiple times.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Automattic/wp-calypso/pull/52139#issuecomment-828960637>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABLHJJ7MH5XD3LUJGTNUXC3TLDYS3ANCNFSM43JRWSNQ>
> .
>

roo2 commented:
> IMO it is, because I worry about the reports we'll get about the the tour appearing for folks it shouldn't appear for, and then it could be exasperating to track down because it'll be based on timing.

I think that doing it in PHP or JS would be equivalent in terms of behaviour.
But yea I do see how putting it in JS is probably more elegant

> How many times would a user repeatedly enter the editor within the first 30 min

It shouldn't matter because once the user has dismissed/completed the modal the closed status will be set in local storage so they won't see it again without opening it from the menu.a

p-jackson commented:
> It shouldn't matter because once the user has dismissed/completed the modal the closed status will be set in local storage so they won't see it again without opening it from the menu.

🤦‍♂️🤦‍♂️🤦‍♂️🤦‍♂️🤦‍♂️

Finally I get it. Yes, it's equivalent isn't it, ignore me 🥲

So yes, php is much more straight forward, and we're still leveraging the local storage as a sort of pseudo-per-site setting. Carry on!

roo2 committed:
added clarifying comments

autumnfjeld reviewed (approved):
Looking Good Eru!   Thanks for allowing for all the discussion on this. I'll write up a P2 post that we can share with HE so they know about this change of showing the Welcome Tour on all new sites and all the discussion that went into it.
